ha/fault_tolerance/rest_api.py

```diff
@@ -0,0 +1,255 @@
+        pass
+
+    @staticmethod
+    def error_response(err: Exception, **kwargs):
+        """
+        Create response object from input exception object.
+        Args:
+            err (Exception): exception object for that error response sent to be client.
+        """
+        # TODO: CORTX-30931 Implement
+        pass
+
+    @staticmethod
+    @web.middleware
+    async def rest_middleware(request, handler):
+        """
+        The 'middleware' coroutine which get called when Request is received to the server.
+        Args:
+            request (web.Request): Request object
+            handler (handler): handler coroutine
+        Returns:
+            Json object: return json object that to be sent to client, ref: CcRestApi.json_response
+        """
+        Log.debug(f"Rest middleware is called: request = {request} handler = {handler}")
+        if CcRestApi.__is_shutting_down:
+            # TODO : CORTX-30931 return json proper response with HTTP status = 503
+            return CcRestApi.json_response("CC REST Server is shutting down", status=503)
+        try:
+            request_id = int(time.time())
+
+            try:
+                await CcRestApi.check_for_unsupported_endpoint(request)
+            except Exception as e:
+                Log.warn(f"Exception: {e}")
+                # TODO:
+                CcRestApi.error_response(Exception("Invalid endpoint."))
+
+            resp = await handler(request)
+
+            if isinstance(resp, web.StreamResponse):
+                return resp
+
+            status = 200
+            if isinstance(resp, Response):
+                status = resp.rc()
+                resp_obj = {'response_body': resp.output(), 'status_code': status}
+                Log.info(f"Response = {resp_obj}")
+
+                if not 200 <= status <= 299:
+                    Log.error(f"Error: ({status}):{resp_obj['response_body']}")
+            else:
+                resp_obj = resp
+                Log.info(f"Response = {resp_obj}")
+
+            return CcRestApi.json_response(resp_obj, status)
+
+        # TODO: CORTX-30931 Handle multiple exceptions, below is the just 1 added
+        except (ConcurrentCancelledError, AsyncioCancelledError) as e:
+            Log.warn(f"Client cancelled call for {request.method} {request.path}")
+            # TODO: CORTX-30931 use CcRequestCancelled return proper response
+            return CcRestApi.json_response(CcRestApi.error_response(Exception("Call cancelled by client"),
+                                            request = request, request_id = request_id), status=499)
+
+    @staticmethod
+    async def _stop():
+        """
+        This function stops the REST Server.
+        cancels all the running tasks, stops site and event loop
+        """
+        CcRestApi.__is_shutting_down = True
+        for task in asyncio.Task.all_tasks():
+            if task != asyncio.Task.current_task():
+                task.cancel()
```

Comment:
But why not to cancel current task as well?
loop.stop will stop it but can be restarted. So, do we want to restart this task again after system is back>

Reply:
Calling cancel() on running Task will cause the Task to throw a CancelledError exception into the wrapped coroutine

Reply:
Ok..